File: service-common/src/main/java/client/ClientSocket.java
package client;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import io.netty.bootstrap.Bootstrap;
import io.netty.buffer.PooledByteBufAllocator;
import io.netty.channel.*;
import io.netty.channel.epoll.Epoll;
import io.netty.channel.epoll.EpollEventLoopGroup;
import io.netty.channel.epoll.EpollSocketChannel;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioSocketChannel;
import io.netty.handler.codec.http.DefaultHttpHeaders;
import io.netty.handler.codec.http.HttpClientCodec;
import io.netty.handler.codec.http.HttpObjectAggregator;
import io.netty.handler.codec.http.websocketx.*;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.val;
import packages.GreetingPackage;
import packages.BridgePackage;
import util.UtilNetty;

import java.net.URI;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

/**
 * @author Рейдж 03.10.2021
 * @project ThePit
 */

@RequiredArgsConstructor
@Getter
@ChannelHandler.Sharable
public class ClientSocket extends SimpleChannelInboundHandler<WebSocketFrame> {

    private static final Class<? extends SocketChannel> CHANNEL_CLASS;
    private static final EventLoopGroup GROUP;

    static {
        boolean epoll;
        try {
            Class.forName("io.netty.channel.epoll.Epoll");
            epoll = !Boolean.getBoolean("cristalix.net.disable-native-transport") && Epoll.isAvailable();
        } catch (ClassNotFoundException ignored) {
            epoll = false;
        }
        if (epoll) {
            CHANNEL_CLASS = EpollSocketChannel.class;
            GROUP = new EpollEventLoopGroup(1);
        } else {
            CHANNEL_CLASS = NioSocketChannel.class;
            GROUP = new NioEventLoopGroup(1);
        }
    }

    private final Cache<String, CompletableFuture<?>> responseCache = CacheBuilder.newBuilder().expireAfterWrite(15, TimeUnit.SECONDS).build();
    private final String host;
    private final int port;
    private final String password;
    private final String serverName;

    private final Map<Class<? extends BridgePackage>, Consumer<?>> handlersMap = new HashMap<>();
    private Channel channel;

    public void connect() {
        new Bootstrap()
                .channel(CHANNEL_CLASS)
                .group(GROUP)
                .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 1000)
                .handler(new ChannelInitializer<Channel>() {
                    @Override
                    protected void initChannel(Channel ch) throws Exception {
                        val config = ch.config();
                        config.setOption(ChannelOption.IP_TOS, 24);
                        config.setAllocator(PooledByteBufAllocator.DEFAULT);
                        config.setOption(ChannelOption.TCP_NODELAY, Boolean.TRUE);
                        config.setOption(ChannelOption.SO_KEEPALIVE, Boolean.TRUE);
                        ch.pipeline()
                                .addLast(new HttpClientCodec())
                                .addLast(new HttpObjectAggregator(6553600))
                                .addLast(new WebSocketClientProtocolHandler(
                                        WebSocketClientHandshakerFactory.newHandshaker(
                                                URI.create("http://" + host + ":" + port + "/"),
                                                WebSocketVersion.V13,
                                                null,
                                                false,
                                                new DefaultHttpHeaders(),
                                                6553600
                                        ),
                                        true
                                ))
                                .addLast(ClientSocket.this);
                    }
                })
                .remoteAddress(host, port)
                .connect()
                .addListener((ChannelFutureListener) future -> {
                    if (future.isSuccess()) {
                        System.out.println("Connection succeeded, bound to: " + (channel = future.channel()));
                    } else {
                        System.out.println("Connection failed");
                        future.cause().printStackTrace();
                        processAutoReconnect();
                    }
                });
    }

    private void sendHandshake() {
        GreetingPackage greetingPackage = new GreetingPackage(password, serverName);
        channel.writeAndFlush(UtilNetty.toFrame(greetingPackage)).addListeners(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE, new ChannelFutureListener() {
            @Override
            public void operationComplete(ChannelFuture future) {
                if (future.isSuccess()) {
                    System.out.println("Handshake completed!");
                } else {
                    System.out.println("Error during handshake");
                    future.cause().printStackTrace();
                    future.channel().close();
                }
            }
        });
    }

    @Override
    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
        if (evt == WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_COMPLETE) {
            sendHandshake();
        }
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, WebSocketFrame msg) throws Exception {
        if (msg instanceof TextWebSocketFrame) {
            val frame = (TextWebSocketFrame) msg;
            BridgePackage pckg = UtilNetty.readFrame(frame);
            CompletableFuture future = responseCache.getIfPresent(pckg.getId());
            if (future != null) {
                responseCache.invalidate(pckg.getId());
                future.complete(pckg);
            }
            Consumer consumer = handlersMap.get(pckg.getClass());
            if (consumer != null)
                consumer.accept(pckg);
        }
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        channel.close();
        channel = null;
        responseCache.invalidateAll();
        processAutoReconnect();
    }

    public <T extends BridgePackage> CompletableFuture<T> writeAndAwaitResponse(T pckg) {
        val future = awaitResponse(pckg);
        write(pckg);
        return future;
    }

    public <T extends BridgePackage> CompletableFuture<T> awaitResponse(T pckg) {
        CompletableFuture<T> future = new CompletableFuture<>();
        responseCache.put(pckg.getId(), future);
        return future;
    }

    public void write(BridgePackage thePitPackage) {
        write(UtilNetty.toFrame(thePitPackage));
    }

    private void write(TextWebSocketFrame frame) {
        channel.writeAndFlush(frame, channel.voidPromise());
    }

    public <T extends BridgePackage> void registerHandler(Class<T> clazz, Consumer<T> consumer) {
        handlersMap.put(clazz, consumer);
    }

    private void processAutoReconnect() {
        System.out.println("Automatically reconnecting in next 1.5 seconds");
        schedule(this::connect, 1500L, TimeUnit.MILLISECONDS);
    }

    public void schedule(Runnable command, long delay, TimeUnit unit) {
        GROUP.schedule(command, delay, unit);
    }
}

File: commons/src/main/kotlin/me/reidj/bridgebuilders/client/ClientSocket.kt
package me.reidj.bridgebuilders.client

import com.google.common.cache.CacheBuilder
import io.netty.bootstrap.Bootstrap
import io.netty.buffer.PooledByteBufAllocator
import io.netty.channel.*
import io.netty.channel.epoll.Epoll
import io.netty.channel.epoll.EpollEventLoopGroup
import io.netty.channel.epoll.EpollSocketChannel
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioSocketChannel
import io.netty.handler.codec.http.DefaultHttpHeaders
import io.netty.handler.codec.http.HttpClientCodec
import io.netty.handler.codec.http.HttpObjectAggregator
import io.netty.handler.codec.http.websocketx.*
import me.reidj.bridgebuilders.`package`.BridgePackage
import me.reidj.bridgebuilders.`package`.GreetingPackage
import me.reidj.bridgebuilders.util.UtilNetty
import java.net.URI
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import java.util.function.Consumer

class ClientSocket(val host: String, val port: Int, private val password: String, private val serverName: String) : SimpleChannelInboundHandler<WebSocketFrame>() {

    private var channelClass: Class<out SocketChannel>?
    private var group: EventLoopGroup

    init {
        val epoll: Boolean = try {
            Class.forName("io.netty.channel.epoll.Epoll")
            !java.lang.Boolean.getBoolean("cristalix.net.disable-native-transport") && Epoll.isAvailable()
        } catch (ignored: ClassNotFoundException) {
            false
        }
        if (epoll) {
            channelClass = EpollSocketChannel::class.java
            group = EpollEventLoopGroup(1)
        } else {
            channelClass = NioSocketChannel::class.java
            group = NioEventLoopGroup(1)
        }
    }

    private val responseCache =
        CacheBuilder.newBuilder().expireAfterWrite(15, TimeUnit.SECONDS).build<String, CompletableFuture<*>>()

    private val handlersMap: MutableMap<Class<out BridgePackage>, Consumer<*>> = mutableMapOf()
    private var channel: Channel? = null

    fun connect() {
        Bootstrap()
            .channel(channelClass)
            .group(group)
            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 1000)
            .handler(object : ChannelInitializer<Channel>() {
                override fun initChannel(ch: Channel) {
                    val config = ch.config()
                    config.setOption(ChannelOption.IP_TOS, 24)
                    config.allocator = PooledByteBufAllocator.DEFAULT
                    config.setOption(ChannelOption.TCP_NODELAY, java.lang.Boolean.TRUE)
                    config.setOption(ChannelOption.SO_KEEPALIVE, java.lang.Boolean.TRUE)
                    ch.pipeline()
                        .addLast(HttpClientCodec())
                        .addLast(HttpObjectAggregator(6553600))
                        .addLast(
                            WebSocketClientProtocolHandler(
                                WebSocketClientHandshakerFactory.newHandshaker(
                                    URI.create("http://$host:$port/"),
                                    WebSocketVersion.V13,
                                    null,
                                    false,
                                    DefaultHttpHeaders(),
                                    6553600
                                ),
                                true
                            )
                        )
                        .addLast(this@ClientSocket)
                }
            })
            .remoteAddress(host, port)
            .connect()
            .addListener(ChannelFutureListener { future: ChannelFuture ->
                if (future.isSuccess) {
                    println("Connection succeeded, bound to: " + future.channel().also { channel = it })
                } else {
                    println("Connection failed")
                    future.cause().printStackTrace()
                    processAutoReconnect()
                }
            })
    }

    private fun sendHandshake() {
        val greetingPackage = GreetingPackage(password, serverName)
        channel!!.writeAndFlush(UtilNetty.toFrame(greetingPackage)).addListeners(
            ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE,
            ChannelFutureListener { future ->
                if (future.isSuccess) {
                    println("Handshake completed!")
                } else {
                    println("Error during handshake")
                    future.cause().printStackTrace()
                    future.channel().close()
                }
            })
    }

    @Throws(Exception::class)
    override fun userEventTriggered(ctx: ChannelHandlerContext, evt: Any) {
        if (evt === WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_COMPLETE) {
            sendHandshake()
        }
    }

    @Throws(Exception::class)
    override fun channelRead0(ctx: ChannelHandlerContext, msg: WebSocketFrame) {
        if (msg is TextWebSocketFrame) {
            val pckg: BridgePackage = UtilNetty.readFrame(msg)
            val future = responseCache.getIfPresent(pckg.id)
            if (future != null) {
                responseCache.invalidate(pckg.id)
                future.complete(pckg as Nothing)
            }
            val consumer = handlersMap[pckg::class.java]
            consumer?.accept(pckg as Nothing)
        }
    }

    override fun channelInactive(ctx: ChannelHandlerContext?) {
        channel!!.close()
        channel = null
        responseCache.invalidateAll()
        processAutoReconnect()
    }

    fun <T : BridgePackage?> writeAndAwaitResponse(pckg: T): CompletableFuture<T> {
        val future = awaitResponse(pckg)
        write(future as BridgePackage)
        return future
    }

    fun <T : BridgePackage?> awaitResponse(pckg: T): CompletableFuture<T> {
        val future = CompletableFuture<T>()
        responseCache.put(pckg!!.id, future)
        return future
    }

    private fun write(thePitPackage: BridgePackage) {
        UtilNetty.toFrame(thePitPackage)?.let { write(it) }
    }

    private fun write(frame: TextWebSocketFrame) {
        channel!!.writeAndFlush(frame, channel!!.voidPromise())
    }

    fun <T : BridgePackage> registerHandler(clazz: Class<T>, consumer: Consumer<T>) {
        handlersMap[clazz] = consumer
    }

    private fun processAutoReconnect() {
        println("Automatically reconnecting in next 1.5 seconds")
        schedule({ connect() }, 1500L, TimeUnit.MILLISECONDS)
    }

    private fun schedule(command: Runnable, delay: Long, unit: TimeUnit?) {
        group.schedule(command, delay, unit)
    }
}